### ISSUE TITLE
Global string search/replace feature

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#104]
[This feature can be found in any standard IDE, i.e. IntelliJ, Flash Builder etc.]

Moonshine needs functionality where user can search through given project (or through all the opened projects on workspace at a time) files for a string value, and replace (optional) the value as well. At present Moonshine has the search/replace feature limited to individual file/editors. 

While developing this, we also need to care that doing such process not to slow-down Moonshine, notably. Investigation on the _worker_ class could be worthy in this case.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine after your last commit I cannot build Moonshine.  IDE doesn't see following files
```
SearchPlugin;
SearchInProjectTree;
```

Did you miss something after ?

**Comment by rat-moonshine** [comment]
Sorry I missed the plugin class to get added. Now added to the Git, please take a pull @piotrzarzycki21 .

**Comment by rat-moonshine** [comment]
Notes to the first commit -

- Added new entry Projects/Search
- Opens a new search in project(s) dialog
- [Works] search text, scope, **Search**
- Opens  new tab with searched patterns between the given 'scope'
- Double-clicking the item (in searched tab) opens to the editor

Things Not included in first commit -

- File name patterns (in search dialog)
- Replace
- Saving 'scope' for consecutive search process
- I wanted auto-height-resize tree, but placed with full height at this moment
- If any UI fine tuning

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I cannot find that entry Projects/Search - It should be in menu Project->Search ?

**Comment by rat-moonshine** [comment]
The worker project is now uploaded to Git.

**Comment by rat-moonshine** [comment]
Change updates:
- Tree is now single component; Projects' data shall be get added/removed to the component
- Progress bar removes when process completes, replaced by search count texts
- Removal of any project from sidebar shall effect to the search results and count text
- Search 'scope' now saves in single application session
- Search tab not requires to close anymore for the next search process
- `File > Close All` now also closes the search result tab

**Comment by rat-moonshine** [comment]
Change updates:
- Regexp search is now working
- File types (pattern) selection

**Comment by rat-moonshine** [comment]
Enhancement requests from Joel, 

> I can't help but think that there may be some situations where I would want to enter this manually, though:
> - Other/custom extensions
> - "build.xml" or "*-app.xml"

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I have a bit bigger expectation from search. It can be expressed as a separate issue if you think that is something not for the 1.8.0.
1) I would like to see in the tree not only information how many strings has been found, but also clickable lines with those strings.
Example:
![image](https://user-images.githubusercontent.com/24554795/34250730-02dbb810-e63e-11e7-8b90-281145ef3cdd.png)

Currently when jump from search to file which have for example 1000 lines and somewhere there is my string - I will be totally lost
2) I think search should have an ability to omit some folders in the project. - This part is an separate feature, because we should have an mechanism for marking some folder as "excluded" in the project tree. 

Currently when I have Royale project and I did build to JavaScript which produce tons of JS files - Search will go through that files. If we have ability to exclude those folder search will be faster. 

Let me know what do you think.

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 we have some other enhancement requests as well to this search feature. Even I was thinking about the option.1 (which we found in Eclipse, too), but we can go for them once the basic criteria for the project search feature meets. 

Please, raise separate issues as enhancement requests once the basic feature is done to this.

**Comment by piotrzarzycki21** [comment]
I just pulled all your changes and I see that I have all the time "Found 0..." even if search found something.
![image](https://user-images.githubusercontent.com/24554795/34252259-2c29216c-e643-11e7-86fc-6b5ee7e38af2.png)

**Comment by rat-moonshine** [comment]
By 'even if search found something' did you mean that "Project > Search" does find something but the count label is coming wrong?

**Comment by piotrzarzycki21** [comment]
This is exactly what I meant.

**Comment by rat-moonshine** [comment]
I confirm, unlike in macOS when tried in Windows I also returned 0 as count value. Looking for the difference.

**Comment by rat-moonshine** [comment]
Counting problem in Windows should be fix now.

**Comment by rat-moonshine** [comment]
Change updates:
- Basic replace
- Search pattern is now editable

**Comment by rat-moonshine** [comment]
Change updates:
- Dual text comparer scroll sync
- File list population in the Preview window
- Text replacement based upon file selections in Preview window

Things not included:
- Completion progress on file selection/replace in Preview window

Known issue:
- Comparer line highlighting (not text/word highlighting) sometime requires scrolling to validate (when changed between files' selection)
- Contents having a few lines sometime do not renders the line highlighting properly

**Comment by piotrzarzycki21** [comment]
@rat-moonshine could you add please some shortcut for global search. I propose on windows: CTRL + SHIFT + F

**Comment by rat-moonshine** [comment]
Change updates:
- Match case scenario handled in the preview section
- File completion progress added in preview section
- Replace by the selected files in the preview section has bug - fixed
- 'Replace' button introduced in the search list tab - should be helpful if user close the replace popup dialog at sometime and want to re-visit

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 I shall look into this.

**Comment by piotrzarzycki21** [comment]
I see following warning in the build, maybe we could avoid it somehow.

```
ws\other\SearchInProjectView.mxml(303): Warning: Data binding will not be able to detect assignments to "isShowReplaceWhenDone".
    [compc]
    [compc]                     <s:Button id="btnReplae"
```

**Comment by rat-moonshine** [comment]
I pushed the fix against it.

**Comment by atinprominic** [comment]
I looked at this feature and at my first glance I recommend below improvements:
 - In search window, highlight the lines which have the matches.
 - In the file opened from search window, highlight the words/ lines which are matched.
 - Provide a shortcut to open Search Dialog.
 - Provide the option to specify project name to search in.

**Comment by rat-moonshine** [comment]
@JoelProminic reported this after some tests last day:
- Create/Open a project with "Test" in the name
- Project > Search
- Enter "Test", and enable "Match Case"
- Click Replace.
- Enter "TestX" as the replace string and click Preview
- Select an as3proj file. Note that there are also some "testMovie" elements that are not matches (because of Match Case)
- Make sure that at least the as3proj file is selected and click "Replace All Selected"
- Close the prompt
- Project > Search
- Enter "TestX". Do not enable "Match Case"
- Click Replace
- Enter "Test" as the replace string and click "Preview"

The above generates the following result where left pane highlighting seems Okay but the right. In right pane it shows some extra highlights of 'Test' (no match case) matches:

![image](https://user-images.githubusercontent.com/24370870/34432214-64594d92-ec9b-11e7-80a3-a4b62f60075b.png)

So far, when I starts thinking about this, I found the right pane is correct from its angle, too. Right pane was highlighting all the 'Test' instances (no match case) without having any know-how what left pane is showing or compared against to.

I'm thinking over this now.

**Comment by rat-moonshine** [comment]
The above problem is still available. However, one fix pushed. There were scenarios when certain lines do not get yellow highlight even after scrolling and having matches, now fixed. Lines should show yellow highlights if they have matches.

**Comment by rat-moonshine** [comment]
Highlighting issue reported by @JoelProminic is fixed now.

**Comment by JoelProminic** [comment]
I noticed some strange behavior when a file has multiple matches.

1. Create a new Royale project called TestPreviewHighlighting
2. Project > Search
3. Enter "test" as the search string, and click "Replace"
4. Enter "testx" as the replacement string, and click "Preview"
5. View the changes for the TestPreviewHighlighting.as3proj file.  Check the "Refactored source" panel.
6. BUG:  The first replacement was applied multiple times
![2018_01_03__testprojecthighlighting_firstmatch](https://user-images.githubusercontent.com/24532731/34544528-66e2c22c-f0ac-11e7-8605-c5c742888fa6.jpg)

7. BUG:  The highlighting is not aligned properly for the other differences (and they also have the above issue to lesser degrees).
![2018_01_03__testprojecthighlighting_othermatches](https://user-images.githubusercontent.com/24532731/34544531-6e6f9b1e-f0ac-11e7-9ae2-b0b2aba3e80b.jpg)

**Comment by JoelProminic** [comment]
I also got a hang when testing the Preview panel:
1. Create/Open a new Royale project called TestPreviewHighlighting
2. Project > Search
3. Enter "test" as the search string, and click "Replace"
4. Enter "testx" as the replacement string, and click "Preview"
5. View the changes for the TestPreviewHighlighting/html-template/$Wrapper.html file.
6. BUG:  Moonshine hangs.

**Comment by rat-moonshine** [comment]
Joel's report follow me to a loop/checking issue (Thank you!) which I didn't tested the same way or overlooked. The case scenario has fixed now, and duplication, wrong caret markings and hang while review tests are fixed now.

Following is a screen captured during my tests, after the fix:
![screenshot](https://user-images.githubusercontent.com/24370870/34553250-771f35e4-f14c-11e7-8e6d-94c207e8eb65.png)

I couldn't able to reproduce Moonshine hang issue during $Wrapper.html file open (to tab) in my tests, though. 

Please, check.

**Comment by rat-moonshine** [comment]
Since the basic file search and replace functionalities has build now, we can push this for a release. It's possible that minor fixes needs here and there, we can continue to work on them. Please, report them here only which are part of the 'basic' functionality.

@piotrzarzycki21 @atinprominic please, raise yours other enhancement requests as separate issues.

I'm closing this for now, please re-open if any critical problem found or report minor basic fixes.
